src/server/decode.rs

```diff
@@ -57,21 +57,28 @@ where
     let method = httparse_req.method;
     let method = method.ok_or_else(|| format_err!("No method found"))?;
 
-    let uri = httparse_req.path;
-    let uri = uri.ok_or_else(|| format_err!("No uri found"))?;
-    let uri = url::Url::parse(&format!("{}{}", addr, uri))?;
+    let path = httparse_req.path;
+    let path = path.ok_or_else(|| format_err!("No uri found"))?;
 
     let version = httparse_req.version;
     let version = version.ok_or_else(|| format_err!("No version found"))?;
     ensure_eq!(version, HTTP_1_1_VERSION, "Unsupported HTTP version");
 
-    let mut req = Request::new(Method::from_str(method)?, uri);
+    let mut req = Request::new(Method::from_str(method)?, url::Url::parse("x:").unwrap());
+
     for header in httparse_req.headers.iter() {
         let name = HeaderName::from_str(header.name)?;
         let value = HeaderValue::from_str(std::str::from_utf8(header.value)?)?;
         req.insert_header(name, value)?;
     }
 
+    let host = req
+        .header(&HOST)
+        .and_then(|header| header.last())
+        .ok_or_else(|| format_err!("Mandatory host header missing"))?;
+
+    *req.url_mut() = url::Url::parse(&format!("http://{}", host.as_str()))?.join(path)?;
```

Comment:
A better alternative would be to use [`Url::set_host`](https://docs.rs/url/2.1.1/url/struct.Url.html#method.set_host) instead.

Reply:
So the idea would be to parse a url with a fake host and then replace it? Why is that preferable?

Reply:
I will test this, but I believe set_host won’t work in all cases because what we’re calling host is actually host[:port] but url::Url::set_host really is just expecting the host part and has a distinct set_port. Is there a reason it’s better to split on : and call set_host and potentially also set_port?

Reply:
Addressed in 0999baf36bef2d76a1d39c48b422d262d3e32bcb but not entirely sure if this is an improvement. On the upside, extracting it to a function improved unit testability